src/lib/crypto/kimchi_backend/gadgets/foreign_field.ml

```diff
@@ -1456,15 +1456,85 @@ let mul (type f) (module Circuit : Snark_intf.Run with type field = f)
   External_checks.append_multi_range_check external_checks
     (quotient_hi_bound, product1_lo, product1_hi_0) ;
 
-  (* Instead of appending external check for compact MRC for remainder,
-   * this is added directly, so that the standard limbs
-   * (remainder0, remainder1, remainder2) are copyable in witness cells *)
   let remainder0, remainder1 =
-    Range_check.compact_multi (module Circuit) remainder01 remainder2
-  in
+    if bound_check_result then (
+      (* Instead of doing the remainder compact-multi-range check externally,
+       * we must insert it directly here, so that the standard limbs
+       * (remainder0, remainder1, remainder2) are copyable in witness cells *)
+      let remainder0, remainder1 =
+        Range_check.compact_multi (module Circuit) remainder01 remainder2
+      in
+      External_checks.append_bound_check external_checks
+        ~do_multi_range_check:false
+        (Element.Standard.of_limbs (remainder0, remainder1, remainder2)) ;
+
+      (remainder0, remainder1) )
+    else
+      (* We do not require a bound check of the result, so save 5.3 rows!
+       *
+       * This happens when the result is already bound-checked outside of this gadget, i.e.,
+       * with some other cvar x, such that x = remainder.
+       *
+       * In this case, we do not want to duplicate the bound check and the compact-range-check.
+       * We only need to make sure that the remainder limbs (remainder0, remainder1, remainder2)
+       * are part of the witness and that remainder0 + 2^L * remainder1 = remainder01.
+       *
+       * We take this approach
+       *     1. Omit adding the bound check to the external-checks (saves 1.8 rows)
+       *     2. Omit doing a compact-multi-range-check on the remainder and do something simpler (saves 3.5 rows)
+       *
+       * Instead of the compact-multi-range-check we do the following
+       *     1. Add copy constrains for remainder0 = x0, remainder1 = x1, remainder2 = x2
+       *        (this is a precondition done outside this gadget)
+       *     2. Compute the decomposition remainder01 = remainder0 + 2^L * remainder1 with a Generic gate
+       *
+       * Details: Since x0, x1, x2 are already bound-checked they are already range-checked < 2^L.
+       * Therefore, in terms of decomposing remainder01, doing (1) and (2) is equivalent to what
+       * the compact-multi-range-check would do. However, it only requires 3 copy constraints and 1
+       * Generic gate, consuming 0.5 rows per double instead of 4.  In the external checks, we also save
+       * an additional generic gate for the high bound computation (0.5 rows) and another range-check
+       * for the high bound range-check (1.3 rows).  This gives a total savings of 5.3 rows per
+       * group double, which is a significant improvement, especially when used in scalar multiplication.
+       *)
+      let remainder0, remainder1 =
```

Comment:
Mmmmmm I am doubting a bit now about this... So if remainder already existed, and thus you had `bound_check_result` set to false, wouldn't this exist duplicate these values? Meaning, it might break wiring between both? Or is this handled somewhere else in the gadget with some asserts between both?

Reply:
Indeed, it duplicates the limbs `remainder0` and `remainder1`, but remember that since they are equal to some other `r0, r1,` as a pre-condition the caller MUST wire them together when this happens and the callers wants to avoid the expensive redundant checks, i.e., as the documentation states
```
* Instead of the compact-multi-range-check we do the following
       *     1. Add copy constrains for remainder0 = x0, remainder1 = x1, remainder2 = x2
       *        (this is a precondition done outside this gadget)
```

IOW, in this usage of "advanced mode" where `bound_check_result` parameter is set to `false` the caller is only doing this because she knows that the result is equal to some other cvar outside the scope of the gadget and she  manually wired those cells / constrained equality between those cvars.

For added clarity, I will update the comment to state

`"* Instead of the compact-multi-range-check the caller MUST do the following"`